### PR TITLE
Add Clickwrap integration

### PR DESCRIPTION
**Pull request by vincentchalamon**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #18
| License       | MIT

### CODE REVIEW

File: features/.env [hunk 1]
```diff
@@ -1,5 +1,7 @@
 DOCUSIGN_INTEGRATION_KEY=c3b2d475-2cbd-47f5-a903-9b3aa0fefe5b
 DOCUSIGN_USER_GUID=4003ab5e-c21f-4738-9469-e7d736e10b90
+DOCUSIGN_CLICKWRAP_ID=a8d45de6-0331-4045-83de-c30da608cedd
+DOCUSIGN_API_ACCOUNT_ID=6623f986-6a54-41b5-bf0b-a07facc4e39d
```
Comment (GregoireHebert) [inline features/.env]:
```suggestion
DOCUSIGN_API_ACCOUNT_ID=05d61a86-31de-4f71-9007-42663b4d4ca0
```

File: features/.env [hunk 2]
```diff
@@ -1,5 +1,7 @@
 DOCUSIGN_INTEGRATION_KEY=c3b2d475-2cbd-47f5-a903-9b3aa0fefe5b
 DOCUSIGN_USER_GUID=4003ab5e-c21f-4738-9469-e7d736e10b90
+DOCUSIGN_CLICKWRAP_ID=a8d45de6-0331-4045-83de-c30da608cedd
```
Comment (GregoireHebert) [inline features/.env]:
```suggestion
DOCUSIGN_CLICKWRAP_ID=91042cd6-8199-489a-926c-fa9527b82142
```

File: features/Kernel.php [hunk 3]
```diff
@@ -130,9 +130,9 @@ protected function configureContainer(ContainerBuilder $c, LoaderInterface $load
                 ],
             ],
             'access_control' => [
-                ['path' => '/embedded', 'roles' => 'IS_AUTHENTICATED_FULLY'],
-                ['path' => '/remote', 'roles' => 'IS_AUTHENTICATED_FULLY'],
-                ['path' => '/', 'roles' => 'IS_AUTHENTICATED_ANONYMOUSLY'],
+                ['path' => '/login', 'roles' => 'IS_AUTHENTICATED_ANONYMOUSLY'],
+                ['path' => '/$', 'roles' => 'IS_AUTHENTICATED_ANONYMOUSLY'],
+                ['path' => '/*', 'roles' => 'IS_AUTHENTICATED_FULLY'],
```
Comment (GregoireHebert) [inline features/Kernel.php]: ? why this change?

File: src/DependencyInjection/DocusignExtension.php [hunk 4]
```diff
@@ -75,12 +76,29 @@ public function load(array $configs, ContainerBuilder $container): void
             ->setPublic(false)
             ->addTag('routing.loader');
 
+        $clickwrapExtensionDefinition = $container->register('docusign.twig.extension.clickwrap', ClickwrapExtension::class)
```
Comment (GregoireHebert) [inline src/DependencyInjection/DocusignExtension.php]: shouldn't we check that twig is loaded? So it's not loaded for nothing.

Comment (GregoireHebert) [reply src/DependencyInjection/DocusignExtension.php]: I suppose a simple class_exist() over Twig_Environment should do.

File: tests/DependencyInjection/DocusignExtensionTest.php [hunk 5]
```diff
@@ -80,6 +81,16 @@ class DocusignExtensionTest extends TestCase
         ],
         'sign_path' => '/foo/sign',
     ]];
+    public const CLICKWRAP_CONFIG = ['docusign' => [
+        'demo' => true,
+        'enable_profiler' => true,
+        'mode' => 'clickwrap',
+        'auth_clickwrap' => [
+            'clickwrap_id' => 'f5de48b3-b323-48a3-bfb1-fa375a7db67b',
```
Comment (GregoireHebert) [inline tests/DependencyInjection/DocusignExtensionTest.php]:
```suggestion
            'clickwrap_id' => '91042cd6-8199-489a-926c-fa9527b82142',
```

File: tests/DependencyInjection/DocusignExtensionTest.php [hunk 6]
```diff
@@ -80,6 +81,16 @@ class DocusignExtensionTest extends TestCase
...
+            'api_account_id' => '797a9f49-cb98-4b0d-85f9-3509bc9cc453',
```
Comment (GregoireHebert) [inline tests/DependencyInjection/DocusignExtensionTest.php]:
```suggestion
            'api_account_id' => '05d61a86-31de-4f71-9007-42663b4d4ca0',
```

File: tests/DependencyInjection/DocusignExtensionTest.php [hunk 7]
```diff
@@ -80,6 +81,16 @@ class DocusignExtensionTest extends TestCase
...
+            'user_guid' => 'c6a6b2f7-40be-4d7d-a4b1-ca438906b77e',
```
Comment (GregoireHebert) [inline tests/DependencyInjection/DocusignExtensionTest.php]:
```suggestion
            'user_guid' => '5ea2b503-51af-4059-8b16-21efc6a0577b',
```

File: tests/DependencyInjection/DocusignExtensionTest.php [hunk 8]
```diff
@@ -310,4 +327,75 @@ public function testItLoadsDemoConfig(): void
...
+
+        $definitionProphecy = $this->prophesize(Definition::class);
+        $containerBuilderProphecy->register('docusign.twig.extension.clickwrap', ClickwrapExtension::class)->shouldBeCalled()->willReturn($definitionProphecy);
+        $definitionProphecy->setPublic(false)->shouldBeCalled()->willReturn($definitionProphecy);
+        $definitionProphecy->addTag('twig.extension')->shouldBeCalled()->willReturn($definitionProphecy);
+        $definitionProphecy->addMethodCall('addConfig', ['default', true, [
+            'environment' => 'https://www.docusign.net',
+            'accountId' => '797a9f49-cb98-4b0d-85f9-3509bc9cc453',
+            'clientUserId' => 'c6a6b2f7-40be-4d7d-a4b1-ca438906b77e',
+            'clickwrapId' => 'f5de48b3-b323-48a3-bfb1-fa375a7db67b',
```
Comment (GregoireHebert) [inline tests/DependencyInjection/DocusignExtensionTest.php]:
```suggestion
            'clickwrapId' => '91042cd6-8199-489a-926c-fa9527b82142',
```

File: tests/DependencyInjection/DocusignExtensionTest.php [hunk 9]
```diff
@@ -310,4 +327,75 @@ public function testItLoadsDemoConfig(): void
...
+            ->willReturn($loaderDefinitionProphecy->reveal())
+            ->shouldBeCalled();
+
+        $definitionProphecy = $this->prophesize(Definition::class);
+        $containerBuilderProphecy->register('docusign.twig.extension.clickwrap', ClickwrapExtension::class)->shouldBeCalled()->willReturn($definitionProphecy);
+        $definitionProphecy->setPublic(false)->shouldBeCalled()->willReturn($definitionProphecy);
+        $definitionProphecy->addTag('twig.extension')->shouldBeCalled()->willReturn($definitionProphecy);
+        $definitionProphecy->addMethodCall('addConfig', ['default', true, [
+            'environment' => 'https://www.docusign.net',
+            'accountId' => '797a9f49-cb98-4b0d-85f9-3509bc9cc453',
```
Comment (GregoireHebert) [inline tests/DependencyInjection/DocusignExtensionTest.php]:
```suggestion
            'accountId' => '05d61a86-31de-4f71-9007-42663b4d4ca0',
```

File: tests/DependencyInjection/DocusignExtensionTest.php [hunk 10]
```diff
@@ -310,4 +327,75 @@ public function testItLoadsDemoConfig(): void
...
+            'clientUserId' => 'c6a6b2f7-40be-4d7d-a4b1-ca438906b77e',
```
Comment (GregoireHebert) [inline tests/DependencyInjection/DocusignExtensionTest.php]:
```suggestion
            'clientUserId' => '5ea2b503-51af-4059-8b16-21efc6a0577b',
```

File: tests/config/docusign.yml [hunk 11]
```diff
@@ -3,24 +3,58 @@ services:
...
+          y_position: 400
+    storage: 'docusign.storage'
+
+  clickwrap:
+    demo: true
+    enable_profiler: true
+    mode: clickwrap
+    auth_clickwrap:
+      api_account_id: '62ec50a7-d431-4818-afe7-11f7d0215492'
+      clickwrap_id: 'ee54b24f-eabf-420d-b628-51edd78669dd'
```
Comment (GregoireHebert) [inline tests/config/docusign.yml]:
```suggestion
      clickwrap_id: '91042cd6-8199-489a-926c-fa9527b82142'
```

File: tests/config/docusign.yml [hunk 12]
```diff
@@ -3,24 +3,58 @@ services:
...
+          x_position: 200
+          y_position: 400
+    storage: 'docusign.storage'
+
+  clickwrap:
+    demo: true
+    enable_profiler: true
+    mode: clickwrap
+    auth_clickwrap:
+      api_account_id: '62ec50a7-d431-4818-afe7-11f7d0215492'
```
Comment (GregoireHebert) [inline tests/config/docusign.yml]:
```suggestion
      api_account_id: '05d61a86-31de-4f71-9007-42663b4d4ca0'
```

File: features/.env (hunk 1 above)
Comment (vincentchalamon) [reply features/.env]: `.env` are fake data, it's not recommended to put real data in it

Comment (vincentchalamon) [reply features/.env]: But you should set those environment variables as secrets in this GitHub settings

File: features/Kernel.php (hunk 3 above)
Comment (vincentchalamon) [reply features/Kernel.php]: To allow any new feature to be logged-in, except for the homepage & the login page

File: src/DependencyInjection/DocusignExtension.php (hunk 4 above)
Comment (vincentchalamon) [reply src/DependencyInjection/DocusignExtension.php]: I don't think it's useful. This is just a simple service, with a tag. If the user uses twig, the tag will be used. Otherwise, the service is available but useless, which is not an error

File: tests/DependencyInjection/DocusignExtensionTest.php (hunk 5 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/DocusignExtensionTest.php]: No, it's tests, we use fake data here

File: tests/DependencyInjection/DocusignExtensionTest.php (hunk 6 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/DocusignExtensionTest.php]: Same

File: tests/DependencyInjection/DocusignExtensionTest.php (hunk 7 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/DocusignExtensionTest.php]: Same

File: tests/DependencyInjection/DocusignExtensionTest.php (hunk 8 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/DocusignExtensionTest.php]: Same

File: tests/DependencyInjection/DocusignExtensionTest.php (hunk 9 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/DocusignExtensionTest.php]: Same

File: tests/DependencyInjection/DocusignExtensionTest.php (hunk 10 above)
Comment (vincentchalamon) [reply tests/DependencyInjection/DocusignExtensionTest.php]: Same

File: tests/config/docusign.yml (hunk 11 above)
Comment (vincentchalamon) [reply tests/config/docusign.yml]: Same

File: tests/config/docusign.yml (hunk 12 above)
Comment (vincentchalamon) [reply tests/config/docusign.yml]: Same
